Scripts/Data/Collection/GameObjectSet.cs

```diff
@@ -0,0 +1,129 @@
+﻿namespace VRTK.Core.Data.Collection
+{
+    using UnityEngine;
+    using UnityEngine.Events;
+    using System;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Collects GameObjects into an unordered set and emits relevant events when the set data changes or is searched.
+    /// </summary>
+    public class GameObjectSet : MonoBehaviour
+    {
+        /// <summary>
+        /// Defines the event for the output <see cref="GameObject"/>.
+        /// </summary>
+        [Serializable]
+        public class UnityEvent : UnityEvent<GameObject>
+        {
+        }
+
+        /// <summary>
+        /// Emitted when a searched element is found.
+        /// </summary>
+        public UnityEvent ElementFound = new UnityEvent();
+        /// <summary>
+        /// Emitted when a searched element is not found.
+        /// </summary>
+        public UnityEvent ElementNotFound = new UnityEvent();
+        /// <summary>
+        /// Emitted when the collection has the last element removed from it and becomes empty.
+        /// </summary>
+        public UnityEvent BecameEmpty = new UnityEvent();
+        /// <summary>
+        /// Emitted when the collection has the first element added to it and becomes populated.
+        /// </summary>
+        public UnityEvent BecamePopulated = new UnityEvent();
+
+        /// <summary>
+        /// The elements currently in the set.
+        /// </summary>
+        public HashSet<GameObject> Elements
+        {
+            get;
+            protected set;
+        } = new HashSet<GameObject>();
+
+
+        /// <summary>
+        /// Adds the given element to the set.
+        /// </summary>
+        /// <param name="element">The element to add.</param>
+        public virtual void AddElement(GameObject element)
+        {
+            if (!isActiveAndEnabled || element == null)
+            {
+                return;
+            }
+
+            if (Elements.Count == 0)
+            {
+                BecamePopulated?.Invoke(element);
+            }
+            Elements.Add(element);
+        }
+
+        /// <summary>
+        /// Removes the given element from the set.
+        /// </summary>
+        /// <param name="element">The element to remove.</param>
+        public virtual void RemoveElement(GameObject element)
+        {
+            if (!isActiveAndEnabled)
+            {
+                return;
+            }
+
+            if (Elements.Remove(element) && Elements.Count == 0)
+            {
+                BecameEmpty?.Invoke(element);
+            }
+        }
+
+        /// <summary>
+        /// Clears all elements from the set.
+        /// </summary>
+        public virtual void ClearElements()
+        {
+            if (!isActiveAndEnabled)
+            {
+                return;
+            }
+
+            Elements.Clear();

```

Comment:
Missing
```c#
BecameEmpty?.Invoke(element);
```

Reply:
I pondered this one.

You can't call `BecameEmpty` because you have no element to invoke with.

The BecameEmpty invokes with the last element that was removed from the set.

Now if you do `Clear` you have no last element (as it's a set so it's not ordered)

Whilst I agree you'd expect BecameEmpty to fire when clear happens, what should be the emitted element? null maybe?

Reply:
Yeah `null` is better than not firing it I'd say. Maybe just document that event data to be the object that was removed and `null` in the case of emptying the whole collection?

If we don't need the element in the event we could remove it (but I assume you had it in for a reason 😛).

Reply:
Yeah you need to know the element that was last emptied out of the set for example when an interactor is finally removed from the set you want to know what GameObject was removed so you can call an event like `Ungrabbed`